controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go

```diff
@@ -156,6 +157,7 @@ func (in *KubeadmControlPlane) ValidateUpdate(old runtime.Object) error {
 		{spec, "version"},
 		{spec, "rolloutAfter"},
 		{spec, "rolloutStrategy", "*"},
+		{status, "version"},
```

Comment:
In general this looks good and logical and I would suggest ignoring the entire status. In my opinion the goal of the update webhook was never to block status changes.

However, I'm slightly confused why we didn't had this issue before. If I understood the situation correctly the current webhook should block all status updates, which would mean that the controller itself cannot function.

Am I / Are we missing something here?

Reply:
Could it be that the webhook is not called for status updates, but only for spec changes? But when those spec changes also contain status changes (somehow through conversion) we validate them here, too?

Reply:
The spec that I apply doesn't have any status, but when I added print statements to the `originalJSON` and `modifiedJSON` in the controller, I could see that one have the version field and the other didn't: https://github.com/kubernetes-sigs/cluster-api/blob/v1.0.2/controlplane/kubeadm/api/v1beta1/kubeadm_control_plane_webhook.go#L166

My understanding is that the conversion brings in the status fields before this comparison happens, which is why I noticed those changes. But I agree with you that for these types of updates, we shouldn't even be considering the values of the status, and would think that ignoring the entire status is better from a controller perspective. I will take a look further tomorrow as well.

Reply:
It looks like the webhook is not called for "pure" status updates.

When I think about it some more, while the webhook was not implemented to prevent status updates the question is why your controller is trying to update the status :).

I wonder if the issue is that you are updating the whole resource with `Update` while you probably don't want to overwrite the status (If I understood the issue correctly). To me it looks like the current webhook prevents status updates via the regular "Update" call, which might be a good thing. Given that status is usually updated via `Status().Patch()` (controller runtime client).

So I think if we allow status updates, we would probably open up the possibility to unintended status updates. I think in your case if the local resources doesn't contain a status it would delete it entirely, which might lead to unintended effects (e.g. the "empty status" bubbling up to the Cluster resource).


We are usually only updating resources with `Patch`.

Reply:
So I actually was able to reproduce this using the `Update` controller-runtime API as well as through `kubectl apply`. I am not suggesting updating status either as that's not what is happening here, but in fact the webhook is retrieving the old resource and the new resource with the status and sees a difference of this `version` field. Are you saying that you were able to reproduce it using `kubectl apply` as well?

Reply:
I wasn't able to reproduce it via `kubectl apply`. I was under the impression that `kubectl apply` only updates the spec of an object (but that could absolutely be wrong).

Would it maybe be possible to get example YAMLs to reproduce? (e.g. based on a CAPD cluster template)

Reply:
I have been testing with vsphere and with a `v1alpha3` version of the `KubeadmControlPlane` spec without the status defined, or any changes to the spec. I even just ran `kubectl get` on the resource to retrieve the `v1alpha3` version of it, saved it to a yaml file, and then reapplied removing the `status` and still see the same error.

I haven't however tried it with docker yet, so let me do some testing there and get back to you on some results.

Reply:
Perfect, thx! I don't doubt it's an issue, I just wasn't able to reproduce (I was literally stepping through the webhook with the debugger :))

Reply:
Hm, which kubectl version are you using?

Reply:
I believe I have `v1.23.3`
```
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.3", GitCommit:"816c97ab8cff8a1c72eccca1026f7820e93e0d25", GitTreeState:"clean", BuildDate:"2022-01-25T21:17:57Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"linux/amd64"}
```

Reply:
Yea I initially wasn't able to reproduce at first but was finally able to, will do a fresh install with docker and try again.

Reply:
For some reason I am not able to reproduce this issue with docker so far. I faced this issue with vsphere where I noticed that the `modifiedJSON` doesn't have the version defined in the status field while the `originalJSON` did, which resulted in this issue happening. I thought it was because of how the conversion logic didn't properly convert the previously applied value, but I am not sure in what situation that would result in the field being missing because the controllers should be always retrieving the status of the v1beta1 version of the object post conversion from v1alpha3. It's super weird to me and will continue trying to investigate why I am running into this.

Reply:
How does the `conversion-data` annotation on the v1alpha3 resource get added and also used from the webhook side for future applies? One thing I randomly tried with docker is removing that annotation using the following command and noticed that I get thrown the following error.

Command run:
```
kubectl edit kubeadmcontrolplane.v1alpha3.controlplane.cluster.x-k8s.io docker-test-2
```

Error thrown:
```
# kubeadmcontrolplanes.controlplane.cluster.x-k8s.io "docker-test-2" was not valid:
# * status.version: Forbidden: cannot be modified
#
```

Reply:
> My assumption is also that it shouldn't differ between docker and vsphere because KubeadmControlPlane is not a provider specific resource.

I agree.

Regarding conversion. I did a session / hackmd about that recently: https://hackmd.io/aYLvrb_jTXiwg_MnpD804w / https://www.youtube.com/watch?v=Mk14N4SelNk


I would expect something like the following flow with the controller:
* controller calls get on the KCP resource
    * APIserver reads v1beta1 KCP resource from etcd
    * APIserver calls the conversion webhook for the KCP resource (configured in the KCP CRD => the KCP controller/webhook)
    * KCP controller is converting the KCP resource to v1alpha3 (and adds the conversion-data annotation)
    * APIserver sends the v1alpha3 resource to the controller

* controller does something locally with the KCP resource and writes it back to the server
    * **Note**: it's important that the controller does not drop the conversion-data annotation
    * APIServer sends a conversion request to the KCP controller, which returns it as v1beta1
    * APIServer sends the KCP resource to mutating (KCP defaulting) and validation webhooks


I"m not 100% sure about the order in the last phase regarding conversion and mutating/validation webhook but I think it should be correct.

The conversion for KCP is implemented here: https://github.com/kubernetes-sigs/cluster-api/blob/6b042456383aa18ba61fc5de5c41901dd7267c60/controlplane/kubeadm/api/v1alpha3/conversion.go#L28-L74

Hm. I never really thought about how conversion affects status to be honest. But based on the code I would say during the conversion from v1alpha3 to v1beta1 we are not restoring the new status.version field (as we do with other fields which have been added with v1beta1) so the error seems logical.

The "current" ressource has .status.version the newly converted one is missing it because the conversion from v1alpha3=>v1beta1 does not restore it.

I'll do another deep-dive later, I'm wondering why I couldn't reproduce it yesterday. But it sounds like I should be able to with `kubectl edit`.

Reply:
Thanks for this information, will review this more thoroughly tomorrow. However if we wanted to fix the conversion to restore the status.version field, does that mean getting it from this `conversion-data` annotation or getting from one of the specs? I talked about this approach in the issue so wanted to get some thoughts there too. Happy to change this PR to accommodate accordingly, that's what I was looking for!

Reply:
It would mean restoring it from the `conversion-data` annotation. That's essentially what is happening in l.41-42 of the linked code above just for `.status.version`. 

I just currently don't understand yet what is actually going on and I want to do that before choosing the solution, but I'll have more info soon :)

Reply:
Okay so I could reproduce it, but I think the issue is essentially that the annotation is not preserved between `get` & `update` in the controller.

Essentially the idea is that when a controller which uses the older version of CAPI gets the old version of the resource but the new version is included in the conversion-data annotation. When that resource is written again (including the annotation) the conversion webhook is able to restore the information.

Concrete example:
KCP added .status.version

controller gets v1alpha3.KCP (with the v1beta1.KCP in the conversion-data annotation)
controller modifies the v1alpha3.KCP
controller writes the v1alpha3.KCP to the server
* if the annotation still exists when writing. status.version can be restored by the conversion webhook and the validating webhook won't complain
* if the annotation is gone .status.version can not be restored and the validating webhook will complain about the change in status
(The version restore is implemented here: https://github.com/kubernetes-sigs/cluster-api/blob/daea2f8d2f5bd8048d8f3b03be42ee8827d0320a/controlplane/kubeadm/api/v1alpha3/conversion.go#L42)

So overall I now think it would be best if the controller is adjusted to preserve the conversion-data between get & update .

If we allow updates to the status the version field would be deleted, which is not ideal.

Reply:
You are referring to the end user's controller to adjust the logic to preserve the conversion-data between get and update and not the CAPI controller right? Just wanted to make sure.

That makes more sense and I would agree with that approach instead of just ignoring. However, if a user were to remove the annotation on accident between those steps, we don't have an approach to pull that information again right? The status is something that is not meant to be updated so I am assuming that is something that can be pulled again, or should be ignored in general anyways. But I understand that we are dealing with the nuances of conversion here so it might not be the best approach.

Reply:
I am also thinking that conversion-data could technically be stale data too and if we are determining the `status.version` from there, that is not completely accurate either as it the responsibility of the controllers to update that field right?

Reply:
> You are referring to the end user's controller to adjust the logic to preserve the conversion-data between get and update and not the CAPI controller right? Just wanted to make sure.

Yes

> That makes more sense and I would agree with that approach instead of just ignoring. However, if a user were to remove the annotation on accident between those steps, we don't have an approach to pull that information again right other than trying to get that resource again?

Yes

> The status is something that is not meant to be updated so I am assuming that is something that can be pulled again or reconciled by the controller again, or should be ignored in general anyways. But I understand that we are dealing with the nuances of conversion here so it might not be the best approach. It seems like conversion-data is something that people need to remember specifically for this resource for updates.

I would say the status is something which is not supposed to be updated by the user. The corresponding (KCP) controller is updating it of course (but via the status endpoint, not via the regular update).

>  It seems like conversion-data is something that people need to remember specifically for this resource for updates.

I think the problem is with the approach to "overwrite" the entire resource. With get + modify struct with the fields you care about + update (or better patch) it should be fine.

> I am also thinking that conversion-data could technically be stale data too and if we are determining the status.version from there, that is not completely accurate either as it the responsibility of the controllers to update that field right?

I think that should be handled by resourceVersion. Essentially you're doing optimistic locking. So when you get+modify+write the resource and the resourceVersion on the server changed in the meantime you get a conflict error on write (and it's expected that you just retry)

Reply:
It seems like `kubectl apply` is a bit different in how it deals with annotations than a `kubectl edit` then. This gives a lot of insight into applying the manifests and it seems like it may be doing what is intended then. Thanks for being patient and explaining it throughout the whole process!

Should we close the PR and the corresponding issue then? I don't think this is something that we need to change something about right?

Reply:
> It seems like kubectl apply is a bit different in how it deals with annotations than a kubectl edit then.

I think the problem is that we used kubectl edit explicitly to delete the annotation :). Any other change should have just worked.

> This gives a lot of insight into applying the manifests and it seems like it may be doing what is intended then.

Yup there is some kind of merging going on. Sooner or later with server-side apply (not sure if it's already enabled per default). 

> Thanks for being patient and explaining it throughout the whole process!

You're welcome! :)

> Should we close the PR and the corresponding issue then? I don't think this is something that we need to change something about right?

That is my take, too. I think we can close both. If there are any further problems, let us know! :)